### PR TITLE
Update to Lumino 2 and JupyterLab 4.0.0a29

### PR DESCRIPTION
**Pull request by afshin**
Resolves https://github.com/jupyter/notebook/issues/6538
Fixes https://github.com/jupyter/notebook/issues/6500
Closes https://github.com/jupyter/notebook/pull/6533

- [x] Revert https://github.com/jupyter/notebook/pull/6499
- [x] Check CodeMirror 6: https://github.com/jupyterlab/jupyterlab/pull/11638
- [x] Check `lsp-extension`: https://github.com/jupyterlab/jupyterlab/pull/12534
- [x] Update Playwright: https://github.com/jupyter/notebook/pull/6533
- [x] Update to TypeScript 4.7 to align with JupyterLab
- [x] Change compilation target from ES2017 to ES2018 to align with JupyterLab
- [ ] Update reference snapshots
- [x] Fix Playwright tests for handling the renaming of files and notebooks, behavior changed with https://github.com/jupyterlab/jupyterlab/pull/12849
- [x] Check other changes from https://github.com/jupyterlab/jupyterlab/releases/tag/v4.0.0a27
- [x] Check other changes from https://github.com/jupyterlab/jupyterlab/releases/tag/v4.0.0a28
- [x] Check other changes from https://github.com/jupyterlab/jupyterlab/releases/tag/v4.0.0a29
- [x] Cell output [auto-scroll](https://github.com/jupyter/notebook/blob/73646334817ffaf3dd6adde7a746a2cc2ff1f9cf/packages/notebook-extension/src/index.ts#L248) now seems to be broken after the update, might be related to https://github.com/jupyterlab/jupyterlab/pull/12695
- [x] Fix cell collapser which does not seem to be working anymore (also no hover style)
- [x] Fix CSS issues:
  - [x] Virtual Notebook changes:
image
  - [x] Side-by-side rendering
image
  - [x] Remove this new horizontal scroll bar:
image
- [x] Make sure shadows are still present: https://github.com/jupyter/notebook/pull/6433
    - The new Virtual Notebook DOM structure requires updating the CSS: https://github.com/jupyterlab/jupyterlab/pull/12554
- [x] Cleanup the notebook CSS

### CODE REVIEW

**Comment by jtpio** [comment]
Thanks @afshin for starting this.

Do you want to take over the lab `4.0.0a27` update as well, which was started in https://github.com/jupyter/notebook/pull/6525 and has some overlap with the current PR? If so happy to close https://github.com/jupyter/notebook/pull/6525 then.

There are a couple of items to have a look listed in the top comment: https://github.com/jupyter/notebook/pull/6525#issue-1352449191

**Comment by afshin** [comment]
@jtpio, sounds like a plan! I may ask you for advice if I run into any headaches 🤕

**Comment by afshin** [comment]
This migration will need to wait for a version of Lumino that includes this bug fix: https://github.com/jupyterlab/lumino/pull/391

**Comment by jtpio** [comment]
Looks like there is a new error coming from `jupyter_ydoc`: https://github.com/jupyter/notebook/actions/runs/3145280672/jobs/5112365511

```
lerna ERR! yarn run build:prod stderr:
Traceback (most recent call last):
  File "/tmp/build-env-sga8vbqf/bin/jupyter-labextension", line 5, in <module>
    from jupyterlab.labextensions import main
  File "/tmp/build-env-sga8vbqf/lib/python3.10/site-packages/jupyterlab/labextensions.py", line 37, in <module>
    from .labapp import LabApp
  File "/tmp/build-env-sga8vbqf/lib/python3.10/site-packages/jupyterlab/labapp.py", line 15, in <module>
    from jupyter_server_ydoc.ydoc import JupyterSQLiteYStore
  File "/tmp/build-env-sga8vbqf/lib/python3.10/site-packages/jupyter_server_ydoc/__init__.py", line 6, in <module>
    from .ydoc import YDocWebSocketHandler
  File "/tmp/build-env-sga8vbqf/lib/python3.10/site-packages/jupyter_server_ydoc/ydoc.py", line 11, in <module>
    from jupyter_ydoc import ydocs as YDOCS  # type: ignore
  File "/tmp/build-env-sga8vbqf/lib/python3.10/site-packages/jupyter_ydoc/__init__.py", line 1, in <module>
    import pkg_resources
```

File: app/package.json [hunk 1]
```diff
@@ -22,76 +22,76 @@
...
+    "@jupyterlab/theme-dark-extension": "~4.0.0-alpha.13",
+    "@jupyterlab/theme-light-extension": "~4.0.0-alpha.13",
+    "@jupyterlab/tooltip": "~4.0.0-alpha.13",
+    "@jupyterlab/tooltip-extension": "~4.0.0-alpha.13",
+    "@jupyterlab/translation": "~4.0.0-alpha.13",
+    "@jupyterlab/translation-extension": "~4.0.0-alpha.13",
+    "@jupyterlab/ui-components": "~4.0.0-alpha.28",
+    "@jupyterlab/vega5-extension": "~4.0.0-alpha.13",
+    "@lumino/algorithm": "^2.0.0-alpha.1",
+    "@lumino/application": "^2.0.0-alpha.1",
```
Comment (jtpio) [inline app/package.json:84]: Looks like we can also update the `@lumino` packages to `^2.0.0-alpha.6` which is the latest version at the moment.

File: pyproject.toml [hunk 2]
```diff
@@ -1,5 +1,5 @@
 [build-system]
-requires = ["hatchling>=1.5", "jupyterlab>=4.0.0a25,<5", "ypy-websocket==0.2"]
+requires = ["hatchling>=1.5", "jupyterlab>=4.0.0a29,<5", "ypy-websocket==0.3.2"]
```
Comment (jtpio) [inline pyproject.toml:2]: It should be possible to drop that one as well:

```suggestion
requires = ["hatchling>=1.5", "jupyterlab>=4.0.0a29,<5"]
```

**Comment by jtpio** [comment]
> **EDIT**: opened [jupyter-server/jupyter_ydoc#58](https://github.com/jupyter-server/jupyter_ydoc/issues/58)

https://github.com/jupyter-server/jupyter_ydoc/issues/58 has been merged and released, restarting the CI checks now.

**Comment by jtpio** [comment]
I updated the top comment with a list of items to check while doing this update: https://github.com/jupyter/notebook/pull/6539#issue-1363038765

**Comment by jtpio** [comment]
FYI @afshin I'll be pushing a couple more updates to the branch directly.

**Comment by jtpio** [comment]
A side-effect of grabbing https://github.com/jupyterlab/jupyterlab/pull/13074 from upstream is the change of the Running tab:

image

I would say it's fine to keep it that way in this PR. We can revisit later whether we want to switch back to the previous look:

image

File: ui-tests/test/smoke.spec.ts [hunk 3]
```diff
@@ -31,19 +31,22 @@ test.describe('Smoke', () => {
...
     runAndAdvance(notebook);
 
     // Enter code in the next cell
-    await notebook.fill(
-      "//div[normalize-space(.)=' ​']/div[1]/textarea",
-      'import this'
-    );
+    await notebook
+      .locator(
+        '.jp-Cell-inputArea >> .cm-editor >> .cm-content[contenteditable="true"]'
```
Comment (jtpio) [inline ui-tests/test/smoke.spec.ts:46]: This seems to be hanging on the Chrome UI tests (not firefox), need to investigate why. Maybe there is the kernel selection dialog grabbing the focus.

**Comment by jtpio** [comment]
Alright CI is all passing now :heavy_check_mark: 

I also did a quick check on Binder and it seems to be looking good there as well.

Found a new issue when opening a JSON file with the code editor, but it seems to be happening in JupyterLab too. Opened https://github.com/jupyterlab/jupyterlab/issues/13170 so it can be fixed upstream.

**Review by jtpio — APPROVED** [review]
Thanks @afshin for starting this update.

In the end it required more changes than initially thought but it's looking good now :sweat_smile: 

Getting this one in and making a new release right after.

**Comment by afshin** [comment]
Thank you @jtpio!

**Pull request merged**